### PR TITLE
Lock jib execution to avoid OverlappingFileLockException in parallel builds

### LINKED ISSUE #11903: Gradle multimodule project + quarkus-container-image-jib: OverlappingFileLockException
**Describe the bug**
I have a gradle + Kotlin multi-module project and try to build container images with the quarkus-container-image-jib extension.  When I run `./gradlew assemble -Dquarkus.container-image.build=true`, I get an `OverlappingFileLockException` (on Windows 10 as well as on the `gradle:6.6-jdk11` docker image). 

Apart from the issues I described in #11900, I can build both images by manually running `./gradlew :serviceA:assemble -Dquarkus.container-image.build=true` and then `./gradlew :serviceB:assemble -Dquarkus.container-image.build=true`. 

Detailed exception: 

```
17:28:34: Executing task 'assemble -Dquarkus.container-image.build=true'...


> Task :serviceA:quarkusGenerateCode
preparing quarkus application
kotlin scripting plugin: project ':serviceA'.compileQuarkus-generated-sourcesKotlin - configuration not found: quarkus-generated-sourcesKotlinScriptDef, the plugin is probably applied by a mistake

> Task :serviceB:quarkusGenerateCode
preparing quarkus application
kotlin scripting plugin: project ':serviceB'.compileQuarkus-generated-sourcesKotlin - configuration not found: quarkus-generated-sourcesKotlinScriptDef, the plugin is probably applied by a mistake

> Task :serviceA:quarkusGenerateCodeTests
preparing quarkus application
kotlin scripting plugin: project ':serviceA'.compileQuarkus-test-generated-sourcesKotlin - configuration not found: quarkus-test-generated-sourcesKotlinScriptDef, the plugin is probably applied by a mistake

> Task :serviceB:quarkusGenerateCodeTests
preparing quarkus application
kotlin scripting plugin: project ':serviceB'.compileQuarkus-test-generated-sourcesKotlin - configuration not found: quarkus-test-generated-sourcesKotlinScriptDef, the plugin is probably applied by a mistake

> Task :serviceB:compileKotlin
> Task :serviceA:compileKotlin
> Task :serviceA:compileJava NO-SOURCE
> Task :serviceB:compileJava NO-SOURCE
> Task :serviceA:processResources
> Task :serviceB:processResources
> Task :serviceA:classes
> Task :serviceB:classes
> Task :serviceA:inspectClassesForKotlinIC
> Task :serviceB:inspectClassesForKotlinIC
> Task :serviceA:jar
> Task :serviceB:jar

> Task :serviceA:quarkusBuild
building quarkus jar

Base image 'fabric8/java-alpine-openjdk11-jre' does not use a specific image digest - build may not be reproducible
Base image 'fabric8/java-alpine-openjdk11-jre' does not use a specific image digest - build may not be reproducible

> Task :serviceB:quarkusBuild
building quarkus jar

> Task :serviceB:quarkusBuild FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':serviceB:quarkusBuild'.
> io.quarkus.builder.BuildException: Build failure: Build failed due to errors
  	[error]: Build step io.quarkus.container.image.jib.deployment.JibProcessor#buildFromJar threw an exception: java.lang.RuntimeException: Unable to create container image
  	at io.quarkus.container.image.jib.deployment.JibProcessor.containerize(JibProcessor.java:153)
  	at io.quarkus.container.image.jib.deployment.JibProcessor.buildFromJar(JibProcessor.java:102)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
  	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:932)
  	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
  	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
  	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
  	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
  	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
  	at java.base/java.lang.Thread.run(Thread.java:830)
  	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
  Caused by: java.util.concurrent.ExecutionException: java.nio.channels.OverlappingFileLockException
  	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:552)
  	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:533)
  	at com.google.common.util.concurrent.FluentFuture$TrustedFuture.get(FluentFuture.java:82)
  	at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$buildImage$8(StepsRunner.java:344)
  	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
  	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
  	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  	at java.base/java.lang.Thread.run(Thread.java:830)
  Caused by: java.nio.channels.OverlappingFileLockException
  	at java.base/sun.nio.ch.FileLockTable.checkList(FileLockTable.java:229)
  	at java.base/sun.nio.ch.FileLockTable.add(FileLockTable.java:123)
  	at java.base/sun.nio.ch.FileChannelImpl.lock(FileChannelImpl.java:1110)
  	at java.base/java.nio.channels.FileChannel.lock(FileChannel.java:1076)
  	at com.google.cloud.tools.jib.filesystem.LockFile.lock(LockFile.java:67)
  	at com.google.cloud.tools.jib.cache.CacheStorageWriter.writeMetadata(CacheStorageWriter.java:315)
  	at com.google.cloud.tools.jib.cache.Cache.writeMetadata(Cache.java:81)
  	at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.pullBaseImage(PullBaseImageStep.java:249)
  	at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call(PullBaseImageStep.java:160)
  	at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call(PullBaseImageStep.java:56)
  	... 6 more


* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 26s
14 actionable tasks: 14 executed
17:29:00: Task execution finished 'assemble -Dquarkus.container-image.build=true'.
```

**To Reproduce**
1. Dowload reproducer project and unzip it [2020-09-04_jib-multimodule-problems.zip](https://github.com/quarkusio/quarkus/files/5175735/2020-09-04_jib-multimodule-problems.zip)
2. Run  `./gradlew assemble -Dquarkus.container-image.build=true`
3. See the error shown above.

**Environment (please complete the following information):**
 - Windows 10 (also happens on `gradle:6.6-jdk11` docker image in my CI builds)
- Gradle
- Kotlin

### PR DESCRIPTION
**Pull request by famod**
It's not pretty (and the locking scope is more coarse than actually needed), but it fixes sproadic `OverlappingFileLockExceptions` when running a parallel Maven build (e.g. `mvn -T2`) in a multi-module project that has at least two modules that build a Quarkus app concurrently with jib.

Reproducer can be found here: https://github.com/quarkusio/quarkus/pull/34873#issuecomment-1655444986
#34873 also has more info of how I ended up with this "solution".

There is a good chance this will also fix #11903, but we need a gradle user to test it.

### CODE REVIEW

**Review by geoand — APPROVED** [review]

File: extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java [hunk 1]
```diff
@@ -311,6 +313,20 @@ private Containerizer createContainerizer(ContainerImageConfig containerImageCon
         return containerizer;
     }
 
+    private JibContainer containerizeUnderLock(JibContainerBuilder jibContainerBuilder, Containerizer containerizer)
+            throws InterruptedException, RegistryException, IOException, CacheDirectoryCreationException, ExecutionException {
+        Class<?> lockObj = getClass();
```
Comment (famod) [inline extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java:318]: I'm on the fence about falling back to just the current class object, which probably never has an effect.
Alternatively, I could only lock if I actually got the class object from parent CL. WDYT?

Comment (aloubyansky) [reply extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java:318]: Thanks @famod for looking into it. Some javadoc explaining the reasons for this change would be good here ;)

Comment (famod) [reply extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java:318]: Yes, of course! I forgot to add a comment myself.

Comment (famod) [reply extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java:323]: I added some javadoc and left the fallback as is.

**Comment by famod** [comment]
So I generated a gradle sample app with jib and it's still working with this change.
The classloaders in Gradle are as follow:
- QuarkusClassLoader:Augmentation Class Loader: PROD@26027e30
- VisitableURLClassLoader(worker-loader)
- FilteringClassLoader(MixInLegacyTypesClassLoader(legacy-mixin-loader))
- MixInLegacyTypesClassLoader(legacy-mixin-loader)
- jdk.internal.loader.ClassLoaders$AppClassLoader@5cb0d902

Which means it's loading QuarkusAugmentor from `VisitableURLClassLoader` and that should be just fine because the next parent `FilteringClassLoader` is not able to load it.
So while I didn't test the actual gradle parallel build scenario that was described in #11903, I'm pretty confident that this change should fix that issue. TBH, I don't think anyone from that issue will report back/try this PR.
Worst thing that can happen is that this change only fixes the Maven case but does nothing for Gradle.

I'm also proposing to backport this to 3.3.

**Review by aloubyansky — APPROVED** [review]

**Pull request merged**